pkg/cloudmeta/aws.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (C) 2024 ScyllaDB
+
+package cloudmeta
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/pkg/errors"
+)
+
+// AWSMetadata is a wrapper around ec2 metadata client.
+type AWSMetadata struct {
+	ec2meta *ec2metadata.EC2Metadata
+}
+
+// NewAWSMetadata is a constructor for  AWSMetadata service.
+func NewAWSMetadata() (*AWSMetadata, error) {
+	session, err := session.NewSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "session.NewSession")
+	}
+	return &AWSMetadata{
+		ec2meta: ec2metadata.New(session),
+	}, nil
+}
+
+// NewAWSMetadataWithEndpoint is a constructor for  AWSMetadata service that allows your to overwrite default metadata endpoint.
+// Might be useful for tests.
+func NewAWSMetadataWithEndpoint(endpoint string) (*AWSMetadata, error) {
+	session, err := session.NewSession()
+	if err != nil {
+		return nil, errors.Wrap(err, "session.NewSession")
+	}
+	cfg := aws.NewConfig()
+	if endpoint != "" {
+		cfg = cfg.WithEndpoint(endpoint)
+	}
+	return &AWSMetadata{
+		ec2meta: ec2metadata.New(session, cfg),
+	}, nil
+}
```

Comment:
I understand you want to introduce some flexibility in terms of choosing different endpoint of AWS service, but why only endpoint ? 
I suggest you to accept `aws.Config` instead of just endpoint.

Besides it's more idiomating to pass whole configuration to constructor in go, instead of small pieces and hardcoding the rest.

Reply:
the only use case I had in mind is to change the endpoint, so AWSMetadata service can be tested against the mock,otherwise I wouldn't expose this method at all. But I see that exposing it makes a lot of confusion. 

Basically until we have CloudMeta service  - all other providers can be private to this package, they are not intended to be used outside of it. Maybe it would better just to make every provider (aws, gcp, azure) as a private service? What do you think?

Reply:
I agree they can be private.
The only thing that this service is expected to return is the provider and instance-type.